compilation error with any GCC version since Feb. 7

```
abundance.cc: In function ‘bool header_find_attribute(const char*, int, const char*, int*, int*)’:
abundance.cc:86:24: error: invalid conversion from ‘const char*’ to ‘char*’ [-fpermissive]
       char * r = strstr(header + i, attribute);
                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
```

--- a/.gitignore
+++ b/.gitignore
@@ -21,3 +21,4 @@ depcomp
 install-sh
 missing
 Makefile.in
+.vscode

--- a/man/vsearch.1
+++ b/man/vsearch.1
@@ -305,7 +305,8 @@ When using \-\-uchime_denovo, the abundance skew is used to
 distinguish in a three-way alignment which sequence is the chimera and
 which are the parents. The assumption is that chimeras appear later in
 the PCR amplification process and are therefore less abundant than
-their parents. The default value is 2.0, which means that the parents
+their parents. For \-\-uchime3_denovo the default value is 16.0. For the other
+commands, the default value is 2.0, which means that the parents
 should be at least 2 times more abundant than their chimera. Any
 positive value equal or greater than 1.0 can be used.
 .TP
@@ -315,7 +316,7 @@ When using \-\-uchimealns, set the width of the three-way alignments
 .TP
 .BI \-\-borderline \0filename
 Output borderline chimeric sequences to \fIfilename\fR, in fasta
-format.  Borderline chimeric sequences are sequences that have a high
+format. Borderline chimeric sequences are sequences that have a high
 enough score but which are not sufficiently different from their
 closest parent.
 .TP
@@ -340,15 +341,18 @@ chimeras, non-chimeras and borderline sequences, using the format
 ';uchime_denovo=\fIfloat\fR;'.
 .TP
 .BI \-\-mindiffs\~ "positive integer"
-Minimum number of differences per segment (default value is 3).
+Minimum number of differences per segment (default value is 3). The parameter is
+ignored with \-\-uchime2_denovo and \-\-uchime3_denovo.
 .TP
 .BI \-\-mindiv \0real
-Minimum divergence from closest parent (default value is 0.8).
+Minimum divergence from closest parent (default value is 0.8). The parameter is
+ignored with \-\-uchime2_denovo and \-\-uchime3_denovo.
 .TP
 .BI \-\-minh \0real
 Minimum score (\fIh\fR). Increasing this value tends to reduce the
 number of false positives and to decrease sensitivity. Default value
-is 0.28, and values ranging from 0.0 to 1.0 included are accepted.
+is 0.28, and values ranging from 0.0 to 1.0 included are accepted. The parameter
+is ignored with \-\-uchime2_denovo and \-\-uchime3_denovo.
 .TP
 .BI \-\-nonchimeras \0filename
 Output non-chimeric sequences to \fIfilename\fR, in fasta
@@ -406,6 +410,19 @@ Detect chimeras present in the fasta-formatted \fIfilename\fR, without
 external references (i.e. \fIde novo\fR). Automatically sort the
 sequences in \fIfilename\fR by decreasing abundance beforehand (see
 the sorting section for details). Multithreading is not supported.
+.TP
+.BI \-\-uchime2_denovo \0filename
+Detect chimeras present in the fasta-formatted \fIfilename\fR, using the 
+UCHIME2 algorithm. This algorithm is designed for denoised amplicons (see
+\-\-cluster_unoise). Automatically sort the sequences in \fIfilename\fR by 
+decreasing abundance beforehand (see the sorting section for details). 
+Multithreading is not supported.
+.TP
+.BI \-\-uchime3_denovo \0filename
+Detect chimeras present in the fasta-formatted \fIfilename\fR, using the 
+UCHIME2 algorithm. The only difference from \-\-uchime2_denovo is that the 
+default minimum abundance skew (\-\-abskew) is set to 16.0 rather
+than 2.0.
 .TP
 .BI \-\-uchime_ref \0filename
 Detect chimeras present in the fasta-formatted \fIfilename\fR by
@@ -2579,9 +2596,10 @@ usearch 7. They are described in the 'Options' section of this
 manual. Here is a short list:
 .RS
 .IP - 2
-alignwidth, borderline, fasta_score (chimera checking)
+uchime2_denovo, uchime3_denovo, alignwidth, borderline, fasta_score (chimera 
+checking)
 .IP -
-cluster_size, clusterout_id, clusterout_sort, profile (clustering)
+cluster_size, cluster_unoise, clusterout_id, clusterout_sort, profile (clustering)
 .IP -
 fasta_width, gzip_decompress, bzip2_decompress (general option)
 .IP -

--- a/src/abundance.cc
+++ b/src/abundance.cc
@@ -83,7 +83,7 @@ bool header_find_attribute(const char * header,
 
   while (i < hlen - alen)
     {
-      char * r = strstr(header + i, attribute);
+      char * r = (char *) strstr(header + i, attribute);
 
       /* no match */
       if (r == NULL)

--- a/src/chimera.cc
+++ b/src/chimera.cc
@@ -785,14 +785,20 @@ int eval_parents(struct chimera_info_s * ci)
       int sumL = best_left_n + best_left_a + best_left_y;
       int sumR = best_right_n + best_right_a + best_right_y;
 
-      if (best_h >= opt_minh)
+      if (opt_uchime2_denovo || opt_uchime3_denovo)
         {
-          status = 3;
-          if ((divdiff >= opt_mindiv) &&
-              (sumL >= opt_mindiffs) &&
-              (sumR >= opt_mindiffs))
-            status = 4;
+          if ((QM == 100.0) && (QT < 100.0))
+          status = 4;
         }
+      else
+        if (best_h >= opt_minh)
+          {
+            status = 3;
+            if ((divdiff >= opt_mindiv) &&
+                (sumL >= opt_mindiffs) &&
+                (sumR >= opt_mindiffs))
+              status = 4;
+          }
 
       /* print alignment */
 
@@ -978,7 +984,7 @@ int eval_parents(struct chimera_info_s * ci)
   0: no parents, non-chimeric
   1: score < 0 (no alignment), non-chimeric
   2: score < minh, non-chimeric
-  3: score >= minh, suspicious
+  3: score >= minh, suspicious -> not available with uchime2_denovo and uchime3_denovo
   4: score >= minh && (divdiff >= opt_mindiv) && ..., chimeric
 */
 
@@ -1279,9 +1285,9 @@ uint64_t chimera_thread_core(struct chimera_info_s * ci)
                 xfree(ci->nwcigar[i]);
                       
               nwcigar = xstrdup(lma.align(ci->query_seq,
-                                         tseq,
-                                         ci->query_len,
-                                         tseqlen));
+                                          tseq,
+                                          ci->query_len,
+                                          tseqlen));
               lma.alignstats(nwcigar,
                              ci->query_seq,
                              tseq,
@@ -1393,7 +1399,7 @@ uint64_t chimera_thread_core(struct chimera_info_s * ci)
             }
           
           /* uchime_denovo: add non-chimeras to db */
-          if (opt_uchime_denovo)
+          if (opt_uchime_denovo || opt_uchime2_denovo || opt_uchime3_denovo)
             dbindex_addsequence(seqno, opt_qmask);
 
           if (opt_nonchimeras)
@@ -1503,10 +1509,10 @@ void chimera()
   opt_self = 1;
   opt_selfid = 1;
 
-  if (opt_uchime_denovo)
+  if (opt_uchime_denovo || opt_uchime2_denovo || opt_uchime3_denovo)
     opt_threads = 1;
 
-  if (opt_uchime_denovo)
+  if (opt_uchime_denovo || opt_uchime2_denovo || opt_uchime3_denovo)
     opt_maxsizeratio = 1.0 / opt_abskew;
 
   tophits = opt_maxaccepts + opt_maxrejects;
@@ -1526,6 +1532,8 @@ void chimera()
   pthread_mutex_init(&mutex_input, NULL);
   pthread_mutex_init(&mutex_output, NULL);
 
+  char * denovo_dbname = NULL;
+
   /* prepare queries / database */
   if (opt_uchime_ref)
     {
@@ -1543,7 +1551,15 @@ void chimera()
     }
   else
     {
-      db_read(opt_uchime_denovo, 0);
+      
+      if (opt_uchime_denovo)
+        denovo_dbname = opt_uchime_denovo;
+      else if (opt_uchime2_denovo)
+        denovo_dbname = opt_uchime2_denovo;
+      else // opt_uchime3_denovo
+        denovo_dbname = opt_uchime3_denovo;
+
+      db_read(denovo_dbname, 0);
 
       if (opt_qmask == MASK_DUST)
         dust_all();
@@ -1601,7 +1617,7 @@ void chimera()
       if (opt_uchime_ref)
         fprintf(fp_log, "%s", opt_uchime_ref);
       else
-        fprintf(fp_log, "%s", opt_uchime_denovo);
+        fprintf(fp_log, "%s", denovo_dbname);
       fprintf(fp_log, ": %d/%u chimeras (%.1f%%)\n",
               chimera_count,
               seqno, 

--- a/src/vsearch.cc
+++ b/src/vsearch.cc
@@ -144,6 +144,8 @@ char * opt_udbinfo;
 char * opt_udbstats;
 char * opt_uc;
 char * opt_uchime_denovo;
+char * opt_uchime2_denovo;
+char * opt_uchime3_denovo;
 char * opt_uchime_ref;
 char * opt_uchimealns;
 char * opt_uchimeout;
@@ -605,7 +607,7 @@ void args_init(int argc, char **argv)
 
   progname = argv[0];
 
-  opt_abskew = 2.0;
+  opt_abskew = -1.0;
   opt_acceptall = 0;
   opt_alignwidth = 80;
   opt_allpairs_global = 0;
@@ -794,6 +796,8 @@ void args_init(int argc, char **argv)
   opt_uc = 0;
   opt_uc_allhits = 0;
   opt_uchime_denovo = 0;
+  opt_uchime2_denovo = 0;
+  opt_uchime3_denovo = 0;
   opt_uchime_ref = 0;
   opt_uchimealns = 0;
   opt_uchimeout = 0;
@@ -1006,6 +1010,8 @@ void args_init(int argc, char **argv)
     {"udbstats",              required_argument, 0, 0 },
     {"cluster_unoise",        required_argument, 0, 0 },
     {"unoise_alpha",          required_argument, 0, 0 },
+    {"uchime2_denovo",        required_argument, 0, 0 },
+    {"uchime3_denovo",        required_argument, 0, 0 },
     { 0, 0, 0, 0 }
   };
 
@@ -1827,6 +1833,14 @@ void args_init(int argc, char **argv)
         case 192:
           opt_unoise_alpha = args_getdouble(optarg);
           break;
+        
+        case 193:
+          opt_uchime2_denovo = optarg;
+          break;
+
+        case 194:
+          opt_uchime3_denovo = optarg;
+          break;
 
         default:
           fatal("Internal error in option parsing");
@@ -1909,6 +1923,11 @@ void args_init(int argc, char **argv)
     commands++;
   if (opt_cluster_unoise)
     commands++;
+  if (opt_uchime2_denovo)
+    commands++;
+  if (opt_uchime3_denovo)
+    commands++;
+
 
   if (commands > 1)
     fatal("More than one command specified");
@@ -2038,6 +2057,15 @@ void args_init(int argc, char **argv)
       else
         opt_minsize = 0;
     }
+      
+  /* set default opt_abskew depending on command */
+  if (opt_abskew < 0.0)
+    {
+      if (opt_uchime3_denovo)
+        opt_abskew = 16.0;
+      else
+        opt_abskew = 2.0;
+    }
 
   /* set default opt_minseqlength depending on command */
 
@@ -2083,6 +2111,8 @@ void args_init(int argc, char **argv)
       & opt_sortbylength,
       & opt_sortbysize,
       & opt_uchime_denovo,
+      & opt_uchime2_denovo,
+      & opt_uchime3_denovo,
       & opt_uchime_ref,
       & opt_udb2fasta,
       & opt_udbinfo,
@@ -2188,15 +2218,17 @@ void cmd_help()
               "\n"
               "Chimera detection\n"
               "  --uchime_denovo FILENAME    detect chimeras de novo\n"
+              "  --uchime2_denovo FILENAME   detect chimeras de novo in a set of denoised amplicons\n"
+              "  --uchime3_denovo FILENAME   detect chimeras de novo in a set of denoised amplicons\n"
               "  --uchime_ref FILENAME       detect chimeras using a reference database\n"
               " Data\n"
               "  --db FILENAME               reference database for --uchime_ref\n"
               " Parameters\n"
-              "  --abskew REAL               min abundance ratio of parent vs chimera (2.0)\n"
+              "  --abskew REAL               min abundance ratio of parent vs chimera (2.0, 16.0 for uchime3)\n"
               "  --dn REAL                   'no' vote pseudo-count (1.4)\n"
-              "  --mindiffs INT              minimum number of differences in segment (3)\n"
-              "  --mindiv REAL               minimum divergence from closest parent (0.8)\n"
-              "  --minh REAL                 minimum score (0.28)\n"
+              "  --mindiffs INT              minimum number of differences in segment (ignored in uchime2/3) (3)\n"
+              "  --mindiv REAL               minimum divergence from closest parent (ignored in uchime2/3) (0.8)\n"
+              "  --minh REAL                 minimum score (ignored in uchime2/3) (0.28)\n"
               "  --sizein                    propagate abundance annotation from input\n"
               "  --self                      exclude identical labels for --uchime_ref\n"
               "  --selfid                    exclude identical sequences for --uchime_ref\n"
@@ -2789,14 +2821,17 @@ void cmd_uchime()
   if (opt_dn <= 0.0)
     fatal("Argument to --dn must be > 0");
 
-  if (opt_mindiffs <= 0)
-    fatal("Argument to --mindiffs must be > 0");
+  if ((!opt_uchime2_denovo) && (!opt_uchime3_denovo))
+  {
+    if (opt_mindiffs <= 0)
+      fatal("Argument to --mindiffs must be > 0");
 
-  if (opt_mindiv <= 0.0)
-    fatal("Argument to --mindiv must be > 0");
+    if (opt_mindiv <= 0.0)
+      fatal("Argument to --mindiv must be > 0");
 
-  if (opt_minh <= 0.0)
-    fatal("Argument to --minh must be > 0");
+    if (opt_minh <= 0.0)
+      fatal("Argument to --minh must be > 0");
+  }
 
 #if 0
   if (opt_abskew <= 1.0)
@@ -2928,7 +2963,7 @@ int main(int argc, char** argv)
     cmd_maskfasta();
   else if (opt_cluster_smallmem || opt_cluster_fast || opt_cluster_size || opt_cluster_unoise)
     cmd_cluster();
-  else if (opt_uchime_denovo || opt_uchime_ref)
+  else if (opt_uchime_denovo || opt_uchime_ref || opt_uchime2_denovo || opt_uchime3_denovo)
     cmd_uchime();
   else if (opt_fastq_chars)
     fastq_chars();

--- a/src/vsearch.h
+++ b/src/vsearch.h
@@ -292,6 +292,8 @@ extern char * opt_udbinfo;
 extern char * opt_udbstats;
 extern char * opt_uc;
 extern char * opt_uchime_denovo;
+extern char * opt_uchime2_denovo;
+extern char * opt_uchime3_denovo;
 extern char * opt_uchime_ref;
 extern char * opt_uchimealns;
 extern char * opt_uchimeout;

